designs/carbon-aware.md

```diff
@@ -0,0 +1,197 @@
++			},
++		}
++	}
+
+	return &Controller{
+		clock:         clk,
+		kubeClient:    kubeClient,
+		cluster:       cluster,
+		provisioner:   provisioner,
+		recorder:      recorder,
+		cloudProvider: cp,
+		lastRun:       map[string]time.Time{},
+		deprovisioners: []Deprovisioner{
+			NewExpiration(clk, kubeClient, cluster, provisioner, recorder),
+			NewDrift(kubeClient, cluster, provisioner, recorder),
+			NewEmptiness(clk),
+			NewEmptyMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
+			NewMultiMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
+			NewSingleMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
+		},
+	}
+}
+```
+</details>
+
+#### Provisioning
+In `karpenter-core`, create a new method `types.go/OrderByCarbonEmissions` and use that in `nodeclaimtemplate.go/ToMachine` and `nodeclaimtemplate.go/ToNodeClaim` instead of `types.go/OrderByPrice` when Carbon Aware is enabled.
+
+In `karpenter`, create a new method `CarbonAwareCreate` in `pkg/providers/instance/instance.go` that is used in `pkg/cloudprovider/cloudprovider.go/Create` instead of `pkg/providers/instance/instance.go/Create` when Carbon Aware is enabled.
+
+#### Considerations
+1. 👍 Current consolidation methods are unaffected.
+1. 👎 There might be copy-paste of code from the original consolidation methods to the carbon aware consolidators.
+
+### Option 2: Use Carbon Aware filtering/sorting methods
+
+#### Consolidation
+Create carbon aware implementations of low-level functions like `filterByPrice`, `filterOutSameType`, `getCandidatePrices`, etc. that is used when Carbon Aware is enabled. Usage of aforementioned functions might assume that it is price that they are getting, but in reality it is data about carbon emissions.
+
+#### Provisioning
+Use same changes to provisioning as in [option 1](#option-1-use-carbon-aware-provisioning-and-concolidation-methods).
+
+#### Considerations
+1. 👍 Less code copy-paste.
+1. 👍 Improvements to original consolidation methods also improve the Carbon Aware feature.
+1. 👎 Has a risk of breaking undocumented invariants.
+1. 👎 Adds complexity to the original consolidation methods.
+
+### Option 3: Override instance price with carbon price (recommended)
+Minimize carbon emissions by defining a price per kgCO₂e and override the instance price with the carbon price (USD/kgCO₂e). Using the `prioritized` launch strategy, carbon emissions will be minimized during provisioning. Consolidation will unknowingly consolidate to minimize carbon emissions.
+
+The carbon price will depend on on `region` and `instanceType` and assume constant resource utilization (e.g. always 80% utilization). The carbon price will be generated in a "hack" and included as consts (same method as used for generating initial pricing[^2]). The carbon price / emission estimates can be updated with new versions.
```

Comment:
Can you clarify how "assume constant resource utilization" concept affects carbon price?

Reply:
This is one place I am constrained in my knowledge about Karpenter, so please correct me if this is incorrect: as far as I know, Karpenter can not provide information about current resource utilization (CPU/RAM) nor about expected utilization for nodes considered to be created from provisioning or consolidation.

To calculate the operational Global Warming Potential (kgCO2e) we must know how much power (W) that an instance ("server") consume. The power consumption is not static, but depends on how large the workload (%) is (e.g. see image). Power consumption is not a linear function of the workload. Therefore, CPU Power Consumption profiles (based on empirical data, aka. server testing) are used to estimate how much power a CPU consume depending on the workload. When calculating the operational carbon emissions we therefore should know what the current/expected workload is to be able to accurately estimate the operational footprint. However, if we do not know the workload (%) we must make an assumption about how large it is to complete the calculation.

I can elaborate further on this in the call today (Wednesday) if you would like, because the calculation is a bit complex to explain in a comment ;) You can read more about the calculation in the [Boavizta API (v.dev) docs here](https://dev.doc.api.boavizta.org/Explanations/components/cpu/#consumption-profile). 

See an illustration of the concept in this picture (AGPL-3.0, Boavizta API): 
![billede](https://github.com/aws/karpenter/assets/16527874/de35f745-d3c7-46a1-9d8e-89a1044e6168)